File: src/types/relationship.ts
export type RelationshipType =
  | 'uses'
  | 'depends_on'
  | 'consumes'
  | 'creates'
  | 'references'
  | 'connected_to'
  | 'similar_to'
  | 'alternative_to'
  | 'parent'
  | 'child'
  | 'replaced_by'
  | 'deprecated_by'
  | 'part_of'
  | 'contains'
  | 'triggers'
  | 'custom'

export const RELATIONSHIP_LABELS: Record<RelationshipType, string> = {
  uses: 'Uses',
  depends_on: 'Depends On',
  consumes: 'Consumes',
  creates: 'Creates',
  references: 'References',
  connected_to: 'Connected To',
  similar_to: 'Similar To',
  alternative_to: 'Alternative To',
  parent: 'Parent',
  child: 'Child',
  replaced_by: 'Replaced By',
  deprecated_by: 'Deprecated By',
  part_of: 'Part Of',
  contains: 'Contains',
  triggers: 'Triggers',
  custom: 'Related To',
}

export interface Relationship {
  id: string
  vaultId: string
  sourceId: string
  targetId: string
  relationshipType: RelationshipType
  label?: string
  strength: number
  description?: string
  bidirectional: boolean
  metadata: Record<string, unknown>
  createdAt: string
}

export interface CreateRelationshipInput {
  vaultId: string
  sourceId: string
  targetId: string
  relationshipType: RelationshipType
  label?: string
  strength?: number
  description?: string
  bidirectional?: boolean
  metadata?: Record<string, unknown>
}

File: src-tauri/src/domain/relationship.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Relationship {
    pub id: Uuid,
    pub vault_id: Uuid,
    pub source_id: Uuid,
    pub target_id: Uuid,
    pub relationship_type: RelationshipType,
    pub label: Option<String>,
    pub strength: f32,         // 0.0 – 1.0
    pub description: Option<String>,
    pub bidirectional: bool,
    pub metadata: serde_json::Value,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RelationshipType {
    Uses,
    DependsOn,
    Consumes,
    Creates,
    References,
    ConnectedTo,
    SimilarTo,
    AlternativeTo,
    Parent,
    Child,
    ReplacedBy,
    DeprecatedBy,
    PartOf,
    Contains,
    Triggers,
    Custom,
}

impl RelationshipType {
    pub fn label(&self) -> &'static str {
        match self {
            Self::Uses => "uses",
            Self::DependsOn => "depends on",
            Self::Consumes => "consumes",
            Self::Creates => "creates",
            Self::References => "references",
            Self::ConnectedTo => "connected to",
            Self::SimilarTo => "similar to",
            Self::AlternativeTo => "alternative to",
            Self::Parent => "parent",
            Self::Child => "child",
            Self::ReplacedBy => "replaced by",
            Self::DeprecatedBy => "deprecated by",
            Self::PartOf => "part of",
            Self::Contains => "contains",
            Self::Triggers => "triggers",
            Self::Custom => "related to",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateRelationshipInput {
    pub vault_id: Uuid,
    pub source_id: Uuid,
    pub target_id: Uuid,
    pub relationship_type: RelationshipType,
    pub label: Option<String>,
    pub strength: Option<f32>,
    pub description: Option<String>,
    pub bidirectional: Option<bool>,
    pub metadata: Option<serde_json::Value>,
}
